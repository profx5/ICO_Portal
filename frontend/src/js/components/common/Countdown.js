import React from 'react';
import moment from 'moment';


class Countdown extends React.Component {

    constructor(props) {
        super(props);
        this.state = {
            days: '00',
            hours: '00',
            minutes: '00',
            seconds: '00'
        };
        this.countdown = null;
    }

    componentDidMount() {
        this.startCountdown();
    }

    componentWillUnmount() {
        clearInterval(this.countdown);
    }
    
    calcDiff = (start, end) => {
        let startTime = + new Date(start);
        let endTime = + new Date(end);

        return (endTime - startTime);
    }

    formatTime = (miliseconds) => {
        let ms = miliseconds;
        let timeArr = moment(new Date(ms)).utcOffset(0).toArray();

        timeArr.shift();
        timeArr.pop();
        
        let time = {
            days:  (timeArr[0] * 30) + timeArr[1],
            hours: timeArr[2],
            minutes: timeArr[3],
            seconds: timeArr[4],
        };
        
        while (time.seconds > 60) {
            time.seconds -= 60;
            time.minutes += 60;
        }
    
        while (time.minutes > 60) {
            time.minutes -= 60;
            time.hours += 24;
        }
    
        while (time.hours > 24) {
            time.hours -= 24;
            time.days += 1;
        }

        return time;
    }

    getTime = () => {
        let {startTime, endTime} = this.props;
        let ms = this.calcDiff(startTime, endTime);

        return this.formatTime(ms);
    }

    startCountdown = () => {
        var time = this.getTime();

        this.countdown = setInterval(() => {

            time.seconds -= 1;
            if (time.seconds < 0) {
                time.minutes -= 1;
                time.seconds = 59;
            }
        
            if (time.minutes <= 0) {
                time.minutes = 59;
                if (time.hours !== 0)  time.hours -= 1;
            }
        
            if (time.hours <= 0 && time.days !== 0) {
                time.hours = 23;
                if (time.days !== 0)  time.days -= 1;
            }
        
            if (time.seconds <= 0 && time.minutes <= 0 && time.hours <= 0 && time.days <= 0) {
                clearInterval(this.countdown);
                console.error('Countdown has been reached it\'s end!');
                this.setState(() => {
                    return {
                        seconds: '00',
                        minutes: '00',
                        hours: '00',
                        days: '00'
                    };
                });
            }

            this.setState(() => {
                return {
                    seconds: time.seconds < 10 ? `0${time.seconds}` : `${time.seconds}`,
                    minutes: time.minutes < 10 ? `0${time.minutes}` : `${time.minutes}`,
                    hours: time.hours < 10 ? `0${time.hours}` : `${time.hours}`,
                    days: time.days < 10 ? `0${time.days}` : `${time.days}`
                }
            })
        }, 1000)
    }

    render() {
        const {days, hours, minutes, seconds} = this.state;
        return this.props.children(days, hours, minutes, seconds);
    }
}

export default Countdown;
