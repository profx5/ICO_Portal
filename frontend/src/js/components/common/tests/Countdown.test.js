import React from 'react';
import {configure, shallow} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

import Countdown from 'js/components/common/Countdown';


configure({ adapter: new Adapter() });

it('should render stateless Countdown', () => {

    const props = {
        startTime: '2018-08-29T15:10:47+00:00',
        endTime: '2018-12-09T15:10:47+00:00'
    }

    const component = shallow(
        <Countdown {...props}>
            {(days, hours, minutes, seconds) => (
                <React.Fragment>{days} {hours} {minutes} {seconds}}</React.Fragment>
            )}
        </Countdown>
    );
    expect(component).toMatchSnapshot();
})
