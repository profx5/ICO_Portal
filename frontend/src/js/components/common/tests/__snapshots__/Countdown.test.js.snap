// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render stateless Countdown 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Countdown
    endTime="2018-12-09T15:10:47+00:00"
    startTime="2018-08-29T15:10:47+00:00"
  >
    [Function]
  </Countdown>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        "00",
        " ",
        "00",
        " ",
        "00",
        " ",
        "00",
        "}",
      ],
    },
    "ref": null,
    "rendered": Array [
      "00",
      " ",
      "00",
      " ",
      "00",
      " ",
      "00",
      "}",
    ],
    "type": Symbol(react.fragment),
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          "00",
          " ",
          "00",
          " ",
          "00",
          " ",
          "00",
          "}",
        ],
      },
      "ref": null,
      "rendered": Array [
        "00",
        " ",
        "00",
        " ",
        "00",
        " ",
        "00",
        "}",
      ],
      "type": Symbol(react.fragment),
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
